### DISCUSSION TITLE
SSHResource logger check error

### DISCUSSION THREAD

**Discussion by shanebusenitz** [dagster-io/dagster#20456] [Q&A]
I recently upgraded from 1.5.4 to 1.6.4 and noticed SSHResource had replaced the `logger` arguement with a private `_logger`. Previously I was setting up the resource like this:

```
from dagster import get_dagster_logger

my_logger = get_dagster_logger()

ssh_connection = SSHResource(
            remote_host="MY_HOST",
            username="USER",
            password="PASSWORD",
            remote_port=22,
            logger=my_logger
            ).get_connection()

# pass ssh_connection into job running op using `ssh_connection`
```

I believe I was getting an error under 1.5.4 with not having a logger passed in however now without the `logger` argument it throws `dagster._check.CheckError: Expected non-None value: None` with the same code (with or without the `logger=my_logger` arg). 

Looking at the resource, that error is referring to:

```
    @property
    def log(self) -> logging.Logger:
        return check.not_none(self._logger)
```

So it would seem `self._logger` is not being set to `context.log`. 

What am I missing to get this working or is this a bug?

**Comment by mlarose** [comment]
The `self._logger` property should be set by the `setup_for_execution` lifecycle hook. But maybe the `log` method is called before that happens? Could you share a code example where this can be reproduced?

**Comment by shanebusenitz** [comment]
Sure, I'm wrapping the resource in my own for configuring ahead of calling it, so that could be it. 
Resource is setup like this:
```
from dagster import ConfigurableResource
from dagster_aws.ssm.resources import ParameterStoreResource
from pydantic import PrivateAttr
from dagster_ssh import SSHResource
import json

class SFTPResource(ConfigurableResource):
    """
    Resource for using SSHResource
    """

    parameter_store: ParameterStoreResource

    parameter_name: str

    _host: str = PrivateAttr()
    _user: str = PrivateAttr()
    _password: str = PrivateAttr()
    _port: str = PrivateAttr()

    def setup_connection(self):

        params_dic = self.parameter_store.fetch_parameters([self.parameter_name])
        info = params_dic.get(self.parameter_name)
        if info is None:
            raise Exception(f"Could not find parameter {self.parameter_name} within AWS SSM Parameter Store")

        d = json.loads(info)

        self._host = d["host"]
        self._user = d["user"]
        self._password = d["password"]
        self._port = d.get("port", 22)

        return SSHResource(
            remote_host=self._host,
            username=self._user ,
            password=self._password,
            remote_port=self._port
            )
```

Op like this:

```
@op()
def my_sftp_sync_op(
    my_sftp: SFTPResource
    ):

    ftp: SFTP = my_sftp.setup_connection()

    ftp.get_connection() # calling this then throws the log error
    
    # do other ftp stuff

```

Job:
```
@job(
    resource_defs={
        "my_sftp": SFTPResource(
        parameter_store=ParameterStoreResource(
            parameter_paths=["path/to/some/conn/params"],
            with_decryption=True
        ),
        parameter_name="path/to/some/conn/params"
        )
    }
)
def my_sftp_sync_job():
    my_sftp_sync_op()
```


Hardcoding SSHResource inside the `@job` `resource_defs` works which means that I'm likely losing the `context` when the resources are initializing right?

**Comment by shanebusenitz** [comment]
### Up to speed
Adding some more context (pun intended) to this and giving a workaround for anyone who might be running into the same issue. 

```
    @property
    def log(self) -> logging.Logger:
        return check.not_none(self._logger)
```
The above property and `logger` argument that was changed a few months ago to private sets the `SSHResource` up to fail unless there is a `context.log` when initialized (calling within the `@job`).  That works great unless you are following the [Handling secrets](https://docs.dagster.io/guides/dagster/using-environment-variables-and-secrets#handling-secrets) docs and need to get a secret param ahead of calling the resource.

### Current workaround
Since I need to get my parameters (`ParameterStoreResource`) ahead of calling `SSHResource` like in the above linked docs guide, the only workaround I have found is build the init resource manually. Which can be done using `build_init_resource_context` which is not ideal. So using the same setup as my previous comment, however now the op looks like this:

```
from dagster import build_init_resource_context
@op()
def my_sftp_sync_op(
    my_sftp: SFTPResource
    ):

    ftp_setup = my_sftp.setup_connection()
    
    ftp_setup.setup_for_execution(build_init_resource_context())

    ftp = ftp_setup.get_connection().open_sftp()
    
    # do other ftp stuff now and it works
```

I realize that is not what `build_init_resource_context()` was intended for, however I've not found a better work around. Open to a better idea or creating a bug issue out of this. I think ideally this uses context.log or get_dagster_logger() based on how it is called.
